### DISCUSSION TITLE
Is there a reason for using uuids?

### DISCUSSION THREAD

**Discussion by flowers9** [endless-sky/endless-sky#9001] [Q&A] [answered]
I noticed that some code tracks uuids for missions, npcs, and ships, but I don't see anything that seems to use the uuids.  I'm working on code that needs to track ships (it creates a copy, but I don't want to create multiple copies if the code gets called multiple times), so I was planning on use the uuid.  However, I found it's using IETF v4 GUIDs, which are essentially a random number, which means there could be collisions (it's not actually unique).  Collisions would be rare (statistically speaking), but I've seen them happen in production code at work, and they can be very annoying.

Does anyone know if there's a reason why uuids were introduced, and if there's any reason they couldn't be replaced with a monotonically increasing counter instead (which would be unique for any given game)?

**Comment by MasterOfGrey** [comment]
@Hurleveur @tibetiroka @quyykk I summon thee all to this here problem.

**Comment by Hurleveur** [comment]
We need them for the take ship PR and identifying things in a unique way (they're overkill if you ask me though)

**Comment by tibetiroka** [comment]
> Does anyone know if there's a reason why uuids were introduced

According to the original PR (#5081):
> In and of itself, this PR is useless, but it will be needed by other PRs that have to uniquely identify ships, missions, or NPCs when loading data back from a file. This allows pointers to ships, NPCs, and missions to be restored.
>
> Used by https://github.com/endless-sky/endless-sky/pull/5070 to load pointers to mission cargo and mission passengers, and https://github.com/endless-sky/endless-sky/pull/5082 to load the mission whose conversation was being displayed.

Essentially, this allows us to reliably refer to and differentiate missions, ships etc. It has been used in PRs like #7505 to fix long-standing bugs.

> which means there could be collisions (it's not actually unique). Collisions would be rare (statistically speaking), but I've seen them happen in production code at work, and they can be very annoying.

There are 121 random bits. I'm not sure how much actual entropy we are working with on average, but as the code seems to rely on the platform's rng, probably enough.

There are 2^121 different UUIDs. For a given N, the chance of not getting a clash in N UUIDs is: (2^121 * (2^121-1) * ... * (2^121 - (N-1)))/((2^121)^N).

Okay, even WolframAlpha cannot give me estimates for when it would reach a given percentage, so here's the wiki page:
> To put these numbers into perspective, the annual risk of a given person being hit by a meteorite is estimated to be one chance in 17 billion,[[5]](https://en.wikipedia.org/w/index.php?title=Universally_unique_identifier&oldid=755882275#cite_note-5) which means the probability is about 0.00000000006 (6 × 10−11), equivalent to the odds of creating a few tens of trillions of UUIDs in a year and having one duplicate. In other words, only after generating 1 billion UUIDs every second for approximately 100 years would the probability of creating a single duplicate reach 50%.

That would be one hell of a save file.

Also, I highly doubt this has come up for you in production unless someone was messing with the RNG, or there were duplicated records, or race conditions not accounted for (such as using the same RNG concurrently, which might not be supported).

**Reply by flowers9** [reply]
>Also, I highly doubt this has come up for you in production

That's what they said. ;)

That code used a sha256 hash rather than a 121 bit random number, and it's possible that one run I couldn't put through their code (because of the duplication) was the only one they'd ever seen, but it happened, and as a result we had data we simply couldn't process.  Until I changed their code to use to use a monotonically increasing number instead, because that's all it actually needed.

Remember, it's not the chance that it happens in one game, but that chance it happens in one of many games by many players over the lifetime of the code. And it'd be easy to change, if there's not something about it being a uuid being necessary (you just give the generator function an incrementor variable that gets saved in the game save file).

**Reply by tibetiroka** [reply, marked as answer]
The chance of this happening is still several orders of magnitudes smaller than the chance of a player copying over content from a previous savefile with the same "monotonically increasing number".

**Reply by tibetiroka** [reply]
> That code used a sha256 hash

There is your problem, then. There are several known attacks based on duplicating SHA hashes. Even if it was accidental, the amount of entropy is massively reduced by just hashing the input data instead of doing actual PRNG.

**Reply by flowers9** [reply]
We generated the data ourselves, so it was totally accidental.  Valid point on the reduced entropy, and I think your reply above is sufficient reason to not look for a "perfect" solution.
